tls/s2n_x509_validator.c

```diff
@@ -357,16 +346,63 @@ S2N_RESULT s2n_x509_validator_validate_cert_chain(struct s2n_x509_validator *val
             S2N_ERR_CERT_MAX_CHAIN_DEPTH_EXCEEDED);
     RESULT_ENSURE(sk_X509_num(validator->cert_chain_from_wire) > 0, S2N_ERR_NO_CERT_FOUND);
 
-    if (!validator->skip_cert_validation) {
-        X509 *leaf = sk_X509_value(validator->cert_chain_from_wire, 0);
-        RESULT_ENSURE_REF(leaf);
-        if (conn->verify_host_fn) {
-            RESULT_ENSURE(s2n_verify_host_information(validator, conn, leaf), S2N_ERR_CERT_UNTRUSTED);
-        }
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_read_leaf_info(struct s2n_connection *conn, uint8_t *cert_chain_in, uint32_t cert_chain_len,
+        struct s2n_pkey *public_key, s2n_pkey_type *pkey_type, s2n_parsed_extensions_list *first_certificate_extensions) {
+    struct s2n_blob cert_chain_blob = {.data = cert_chain_in, .size = cert_chain_len};
+    DEFER_CLEANUP(struct s2n_stuffer cert_chain_in_stuffer = {0}, s2n_stuffer_free);
+
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&cert_chain_in_stuffer, &cert_chain_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_write(&cert_chain_in_stuffer, &cert_chain_blob));
 
-        RESULT_GUARD_OSSL(X509_STORE_CTX_init(validator->store_ctx, validator->trust_store->trust_store, leaf,
-                validator->cert_chain_from_wire), S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+    struct s2n_blob asn1_cert = {0};
+    RESULT_GUARD(s2n_read_asn1_cert(&cert_chain_in_stuffer, &asn1_cert));
 
+    RESULT_ENSURE(s2n_asn1der_to_public_key_and_type(public_key, pkey_type, &asn1_cert) == 0,
+                  S2N_ERR_CERT_UNTRUSTED);
+
+    /* certificate extensions is a field in TLS 1.3 - https://tools.ietf.org/html/rfc8446#section-4.4.2 */
+    if (conn->actual_protocol_version >= S2N_TLS13) {
+        s2n_parsed_extensions_list parsed_extensions_list = { 0 };
+        RESULT_GUARD_POSIX(s2n_extension_list_parse(&cert_chain_in_stuffer, &parsed_extensions_list));
+
+        /* RFC 8446: if an extension applies to the entire chain, it SHOULD be included in the first CertificateEntry */
+        *first_certificate_extensions = parsed_extensions_list;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_x509_validator_validate_cert_chain(struct s2n_x509_validator *validator, struct s2n_connection *conn,
+        uint8_t *cert_chain_in, uint32_t cert_chain_len, s2n_pkey_type *pkey_type, struct s2n_pkey *public_key_out) {
+    switch (validator->state) {
+        case INIT:
+            break;
+        default:
+            RESULT_BAIL(S2N_ERR_INVALID_CERT_STATE);
+    }
```

Comment:
```suggestion
    POSIX_ENSURE(validator->state == INIT, S2N_ERR_INVALID_CERT_STATE);
```

Reply:
Fixed in 5f7443e6449341641b73f84dd4a574734bcd1982!